[tools] Add compilation database generator for Bazel

nettle opened the pull request:
The tool takes bazel build command as an input parameter
and executes Bazel Action Query (bazel aquery)
to extract all compile commands then saves them
to compile_commands.json file.

csordasmarton commented on tools/bazel/compile_commands/compile_commands.py:
```diff
@@ -0,0 +1,255 @@
+    return options
+
+
+def split_to_list(arguments):
+    """
+    Split argument string to list
+    """
+    if type(arguments) is list:
+        return arguments
+    else:
+        return shlex.split(arguments)
```
There is no need for this function. Use `shlex.split` where it is necessary.

nettle replied:
Yes, I agree that we can use `shlex.split` however `split_to_list` is more convenient and safer.
I would keep it if you dont mind.

csordasmarton commented on tools/bazel/compile_commands/compile_commands.py:
```diff
@@ -0,0 +1,255 @@
+    if not output_file_name:
+        logging.error("No output file is specified")
+        return False
+
+    bazel_info = split_to_list(build_command)[0] + " info "
```
```suggestion
    cmd = shlex.split(build_command)
    bazel_info = cmd[0] + " info "
```

nettle replied:
Yes, I agree that we can use `shlex.split` however `split_to_list` is more convenient and safer.
I would keep it if you dont mind.

csordasmarton commented on tools/bazel/compile_commands/compile_commands.py:
```diff
@@ -0,0 +1,255 @@
+    logging.info("Bazel work dir:      %s", bazel_work_dir)
+
+    logging.info("Bazel build options: %s", build_command)
+    data = bazel_aquery(build_command)
```
Use the `cmd` here, so this function will expect a list instead of a string. And there will be no need to call the split function on the argument.

nettle replied:
Yes, I agree that we can use `shlex.split` however `split_to_list` is more convenient and safer.
I would keep it if you dont mind.

csordasmarton commented on tools/bazel/compile_commands/compile_commands.py:
```diff
@@ -0,0 +1,255 @@
+    """
+    Run shell command
+    """
+    cmd = split_to_list(command)
```
On the call side pass the command as a list of string and there will be no need to split the argument here.

nettle replied:
Yes, I agree that we can use `shlex.split` however `split_to_list` is more convenient and safer.
I would keep it if you dont mind.

csordasmarton commented on tools/bazel/compile_commands/compile_commands.py:
```diff
@@ -0,0 +1,255 @@
+    logging.debug("Command: %s", command)
+
+    out = run_command(command)
+    if out:
+        data = json.loads(out)
+        return data
+    else:
+        return False
```
Now the function may return with `None` (see the code above L101), a dictionary or a Boolean (False) value. Couldn't we return here with None if the output is empty?
```suggestion
    if out:
        return json.loads(out)
```

nettle replied:
Done

csordasmarton commented on tools/bazel/compile_commands/compile_commands.py:
```diff
@@ -0,0 +1,255 @@
+
+def get_full_path(file, directories):
+    """
+    Return full path for a file and try to prepend directories
```
This function will return not only with the full path but the parent diectory too. It would be better to return only with the full path or None and use the `os.path.dirname` to get the parent directory for the file if its not None on the call side.

nettle replied:
Done:

- Changed the name and docstring of the function since `get_full_path` is meisleading

csordasmarton commented:
Please also fix the commit message: `[tols]` -> `[tools]`

nettle commented on tools/bazel/compile_commands/compile_commands.py:
```diff
@@ -0,0 +1,255 @@
+#!/usr/bin/env python3
```
Thanks for reviewing @csordasmarton!
I'll check and update the patch as soon as I have time.

Meanwhile I have a few questions:
1) Regarding the module name I think `bazel_to_compile_commands` has kind of wrong meaning, it is not a converter.
  Probably `bazel_compile_commands` would be better. Agree?
2) I put this module into `bazel` folder, is it OK?
3) How can find information about verification infrastructure you use?
Are there any guidelines? Check list?

csordasmarton replied:
@nettle
1. Yes, you are right, `bazel_compile_commands` will be good.
2. You can put this into the `bazel` folder (`tools/bazel/bazel_compile_commands/bazel_compile_commands.py`)
3. No, there isn't any guidline for this. This is how we built our others tools (https://github.com/Ericsson/codechecker/tree/master/tools, https://github.com/Ericsson/codechecker/tree/master/analyzer/tools). And it is good to be consistent in the whole repository.

nettle replied:
Done:

- Renamed to `bazel_compile_commands`
- Using `bazel` folder
- Added copying to `bin` folder

nettle commented on tools/bazel/tests/Makefile:
```diff
@@ -0,0 +1,37 @@
+# Nose test runner configuration options.
+NOSECFG = --config .noserc
+
+test: pycodestyle pylint test_unit
```
Done:

- Added to Github Action workflow

nettle commented on tools/bazel/setup.py:
```diff
@@ -0,0 +1,34 @@
+    long_description = fh.read()
+
+setuptools.setup(
+    name="bazel-compile-commands",
```
Done

nettle commented on tools/bazel/compile_commands/compile_commands.py:
```diff
@@ -0,0 +1,255 @@
+    parser.add_argument("-v", "--verbosity",
+                        default=0,
+                        action="count",
+                        help="increase output verbosity (e.g., -v or -vv)")
```
Done:

- Verbosity option (`-v` or `--verbose`) works the same way as in `tu_collector`

nettle commented on tools/bazel/README.md:
```diff
@@ -0,0 +1,40 @@
+make venv
+source $PWD/venv/bin/activate
+
+# Build and install plist-to-html package.
```
Done

nettle commented on tools/bazel/tests/Makefile:
```diff
@@ -0,0 +1,37 @@
+UNIT_TEST_CMD = $(REPO_ROOT) \
+  nosetests $(NOSECFG) tests/unit
+
+test_unit: dep
```
Done

nettle commented on tools/bazel/tests/Makefile:
```diff
@@ -0,0 +1,37 @@
+test_unit: dep
+	$(UNIT_TEST_CMD)
+
+test_unit_in_env: venv_dev dep
```
Done

nettle commented on tools/bazel/tests/unit/test_compile_commands.py:
```diff
@@ -0,0 +1,24 @@
+    def setUp(self):
+        pass
+
+    def test_something(self):
```
Done:

- Added simple Bazel example to `tests/projects`
- Added a few simple tests (more tests will be added later)
- Using `@unittest.skipUnless(shutil.which("bazel"), "'bazel' binary not found!")` decorator

nettle commented on tools/bazel/README.md:
```diff
@@ -0,0 +1,40 @@
+# Bazel compilation database generator
+`bazel-compile-commands` is a python tool which generates compilation database
+(`compile_commands.json` file) from Bazel build command.
```
Done:

- Added a few more words about `bazel aquery` approach

Not Done:

- I guess we should change the documentation a bit later, when this script is confirmed as 100% reliable solution

nettle commented:
Hi @csordasmarton,

I've updated the patch upon comments.
Hopefully now it has all required infrastructural parts :)
Please note that a few test cases I've added are just a placeholder and more tests will be added later.
The patch is also missing some functional improvements which I plan to add later.
I guess it is important to move consecutively, so I consider this patch as earlier version of compile commands generator for Bazel, it will be evolving.

nettle commented on tools/bazel/.noserc:
```diff
@@ -0,0 +1,13 @@
+stop=1
+
+# do not capture stdout
+nocapture=1
```
Done

nettle commented on tools/bazel/tests/unit/test_compile_commands.py:
```diff
@@ -0,0 +1,53 @@
+    @unittest.skipUnless(shutil.which("bazel"), "bazel binary not found!")
+    def test_run_bazel(self):
+        """ Test run() function with bazel build ... """
+        self.assertTrue(bazel_compile_commands.run(
+            "bazel build ...", self.TEST_COMPILE_COMMANDS_FILENAME))
```
Good idea!
Done

nettle commented on tools/bazel/tests/unit/test_compile_commands.py:
```diff
@@ -0,0 +1,53 @@
+        """ Test run() function with no arguments. """
+        self.assertFalse(bazel_compile_commands.run("", ""))
+
+    @unittest.skipUnless(shutil.which("bazel"), "bazel binary not found!")
```
Done

nettle commented on tools/bazel/bazel_compile_commands/bazel_compile_commands.py:
```diff
@@ -0,0 +1,269 @@
+    parser.add_argument("-b", "--build",
+                        help="bazel build command arguments")
+    parser.add_argument("-v", "--verbosity",
+                        default=0,
```
OK
Done

nettle commented:
Hi @csordasmarton,
The patch is updated, please take a look!

csordasmarton reviewed (approved):
It's LGTM! Thank you for your time that you implemented it :blush: 

Do you plan to add anything to this PR or can we merge it?

nettle commented:
Thanks for review @csordasmarton!
I think this patch is good to go now.
I will prepare some changes and submit them later as a new PR, if you dont mind.
